lib/px/service/firehose/base_request.rb

```diff
@@ -0,0 +1,118 @@
+require "digest"
+require "redis"
+require "aws-sdk"
+require "px-service-legacy-client"
+require "circuit_breaker"
+
+module Px::Service::Firehose
+  class BaseRequest
+    include Px::Service::Legacy::Client::CircuitBreaker
+
+    DEFAULT_PUT_RATE = 0.25
+    MAX_QUEUE_LENGTH = 1000
+
+    attr_accessor :stream, :credentials
+    attr_reader :firehose, :buffer
+
+    # Circuit breaker configuration
+    circuit_handler do |handler|
+      handler.logger = nil # Or the configured logger
+      handler.failure_threshold = 5
+      handler.failure_timeout = 7
+      handler.invocation_timeout = 10
+      handler.excluded_exceptions = [Px::Service::ServiceRequestError]
+    end
+
+    def initialize
+      @firehose = Aws::Firehose::Client.new(credentials: (@credentials || Px::Service::Firehose.config.credentials), region: Px::Service::Firehose.config.region)
+      @redis = Px::Service::Firehose.config.redis
+      @dev_queue_key = Px::Service::Firehose.config.dev_queue_key
+      @last_send = Time.now
+      @last_throughput_exceeded = nil
+
+      @buffer = []
+      @semaphore = Mutex.new
+    end
+
+    ##
+    # Check if buffer should be flushed and sent to Firehose.
+    def flush_records
+      # clear out nil value in buffer
+      # TODO: fix and figure out why this is happening
+      @buffer = @buffer.compact
+      if @buffer.present? && can_flush?
+        if Px::Service::Firehose.config.dev_mode && @redis && @dev_queue_key
```

Comment:
Like I mentioned in the other PR, if there's any way we can get away from conditions like `if (in dev mode) do {something} else {something else}` it would make the code less cluttered. Doesn't have to be done ASAP but if it can be done it would be better IMO.